src/test/webapp/script/postcardgen_test.js

```diff
@@ -0,0 +1,54 @@
+
+describe('Postcard - check constructor', () => {
+  it('should have title', () => {
+    const title = 'Happy birthday!';
+    let postcard = new Postcard({title});
+    expect(postcard.title).toBe(title);
+  });
+
+  it('should have message', () => {
+    const message = 'Another adventure filled year awaits you.';
+    let postcard = new Postcard({message});
+    expect(postcard.message).toBe(message);
+  });
+
+  it('should have image', () => {
+    const imgUrl = 'https://tse2.mm.bing.net/th/id/OIP.geCKcqpyVwMD6EJuAT3lVQHaEK?w=333&h=187&c=7&o=5&pid=1.7';
+    let postcard = new Postcard({imgUrl});
+    expect(postcard.img).toBe(imgUrl);
+  });
+});
+
+describe('Postcard - generate elements', () => {
+  it('should generate title element', () => {
+    const title = 'Happy birthday!';
+    let postcard = new Postcard({title});
+    expect(postcard.addTitleElem().outerHTML).toBe('<div class="pcard-title">' + title + '</div>');
```

Comment:
Is it really necessary to test against exact HTML? This might not be as robust to insignificant and inconsequential changes in the implementation (like double quotes versus single quotes) while at the same time you don't get a guarantee that the HTML is going to be rendered in the way to the user that you want.

Also such tests can be close to _change detector tests_, tests that nearly implement the same logic as the underlying implementation. The difference here is that you are stringing together the HTML in the test whereas in the implementation you're doing that by using javascript's interface to the DOM (you could string together elements and let javascript deal with creating the DOM, so the difference is not that big).

https://testing.googleblog.com/2015/01/testing-on-toilet-change-detector-tests.html describes the idea of a change detector test.

My suggestion here is to e.g. just make sure that the message appears in the rendered text which is robust to changes in whitespace and quotes etc and how "outerHTML" represents the DOM as html.

`expect(postcard.addMessageElem() ...).toContain(message);` 

without knowing the exact way on how to express this in the test framework.